corehq/celery.py

```diff
@@ -29,24 +29,50 @@
     pass
 
 
+class TimeToStartTimer(object):
+    def __init__(self, task_id):
+        self.task_id = task_id
+
+    @property
+    def _cache_key(self):
+        return 'task.{}.time_sent'.format(self.task_id)
+
+    def start_timing(self):
+        cache.set(self._cache_key, datetime.datetime.utcnow(), timeout=3 * 24 * 60 * 60)
+
+    def stop_and_pop_timing(self):
+        """
+        Return timedelta since running start_timing
+
+        Only the first call to stop_and_pop_timing will return a timedelta;
+        subsequent calls will return None until the next time start_timing is called.
+
+        This helps avoid double-recording timings (for example when a task is retried).
+        """
+        try:
+            return datetime.datetime.utcnow() - cache.get(self._cache_key)
+        except TypeError:
```

Comment:
This seems like an overly broad way to detect that `cache.get(self._cache_key)` returned `None`. What do you think of this?
```py
time_sent = cache.get(self._cache_key)
if time_sent is None:
    return None  # or maybe `raise TimingNotAvailable`
cache.delete(self._cache_key)
return datetime.datetime.utcnow() - time_sent
```
That way if a value with a bad type ends up stored at that key, it will cause an error rather than being swallowed silently.

Another thing to consider is raising an exception rather than returning `None`. Then the calling code does not need to do an error-code check on the return value of this method; instead it can catch the exception and handle it appropriately. This is a style recommended by the _Clean Code_ book we're reading in book club. It has a whole section entitled "Don't Return Null"—recommended reading since it's an anti-pattern we do all over the place.

Reply:
OK I agree with all of this. I think I'll go with the approach of raising an exception in a followup PR like you suggested